fdbserver/TagPartitionedLogSystem.actor.cpp

```diff
@@ -2041,40 +2047,128 @@ ACTOR Future<Void> TagPartitionedLogSystem::getDurableVersionChanged(LogLockInfo
+                                 std::vector<std::tuple<int, std::vector<TLogLockResult>>>& logGroupResults,
                                  Version minEnd) {
+	std::vector<std::vector<int>> tLogLocIds;
+	int maxTLogLocId = std::numeric_limits<int>::min();
+	getTLogLocIds(logServers, logGroupResults, tLogLocIds, maxTLogLocId);
+	int bsSize = maxTLogLocId + 1; // bitset size, used below
+
+	std::vector<Version> RVs(maxTLogLocId + 1, minEnd); // recovery versions of various tLogs
+
+	int tLogGroupIdx = 0;
 	Version minLogGroup = std::numeric_limits<Version>::max();
 	for (auto& logGroupResult : logGroupResults) {
-		std::unordered_map<Version, int> versionRepCount;
-		std::map<Version, int> versionTLogCount;
+		int tLogIdx = 0;
+		boost::dynamic_bitset<> availableTLogs(bsSize);
+		// version -> tLogs (that are avaiable) that have received the version
+		std::unordered_map<Version, boost::dynamic_bitset<>> versionAvailableTLogs;
+		// version -> all tLogs that the version was sent to (by the commit proxy)
+		std::map<Version, boost::dynamic_bitset<>> versionAllTLogs;
+		// version -> prevVersion (that was given out by the sequencer) map
+		std::map<Version, Version> prevVersionMap;
 		int replicationFactor = std::get<0>(logGroupResult);
 		for (auto& tLogResult : std::get<1>(logGroupResult)) {
+			int tLogLocId = tLogLocIds[tLogGroupIdx][tLogIdx];
+			availableTLogs.set(tLogLocId);
 			bool logGroupCandidate = false;
 			for (auto& unknownCommittedVersion : tLogResult.unknownCommittedVersions) {
 				Version k = std::get<0>(unknownCommittedVersion);
 				if (k > minEnd) {
-					versionRepCount[k]++;
-					versionTLogCount[k] = std::get<1>(unknownCommittedVersion);
+					if (versionAvailableTLogs[k].empty()) {
+						versionAvailableTLogs[k].resize(bsSize);
+					}
+					versionAvailableTLogs[k].set(tLogLocId);
+					prevVersionMap[k] = std::get<1>(unknownCommittedVersion);
+					if (versionAllTLogs[k].empty()) {
+						versionAllTLogs[k].resize(bsSize);
+					}
+					populateBitset(versionAllTLogs[k], std::get<2>(unknownCommittedVersion));
 					logGroupCandidate = true;
 				}
 			}
 			if (!logGroupCandidate) {
 				return minEnd;
 			}
+			tLogIdx++;
 		}
 		Version minTLogs = minEnd;
-		for (auto const& [version, tLogCount] : versionTLogCount) {
-			if (versionRepCount[version] >= tLogCount - replicationFactor + 1) {
-				minTLogs = version;
+		Version prevVersion = minEnd;
+		for (auto const& [version, tLogs] : versionAllTLogs) {
+			if (prevVersion == minEnd || prevVersion == prevVersionMap[version]) {
+				// This version is not recoverable if there is a log server (LS) such that:
+				// - the commit proxy sent this version to LS (i.e., LS is present in "versionAllTLogs[version]")
+				// - LS is available (i.e., LS is present in "availableTLogs")
+				// - LS didn't receive this version (i.e., LS is not present in "versionAvailableTLogs[version]")
+				if (((tLogs & availableTLogs) & ~versionAvailableTLogs[version]).none()) {
```

Comment:
We can do "-", instead of doing "& ~" here. Please let me know if that is easier to read, thanks!